src/main/java/net/snowflake/ingest/streaming/internal/CustomCompressionCodec.java

```diff
@@ -0,0 +1,118 @@
+package net.snowflake.ingest.streaming.internal;
+
+import org.apache.arrow.compression.Lz4CompressionCodec;
+import org.apache.arrow.compression.ZstdCompressionCodec;
+import org.apache.arrow.memory.ArrowBuf;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.util.MemoryUtil;
+import org.apache.arrow.vector.compression.CompressionCodec;
+import org.apache.arrow.vector.compression.CompressionUtil;
+
+/**
+ * This is custom wrapper on top of {@link CompressionCodec} implementations.
+ *
+ * <p>The main stream {@link CompressionCodec} implementations are based on {@link
+ * org.apache.arrow.vector.compression.AbstractCompressionCodec}. This implementation is similar to
+ * it with some tweaks.
+ *
+ * <p>The reason to have this custom implementation is to fix certain issues with mainstream codecs
+ * that lead to buffer reference counting problems due to creation of new compressed buffers and
+ * releasing uncompressed. The release of uncompressed buffer from the original vectors to write
+ * into file results in double release when the original vector gets released as usual.
```

Comment:
Do you think we have a better way to resolve this issue? My concern is that we have a few assumptions on how the arrow library would work and that might change in the future